File: gl/src/main/java/com/ryanharter/android/gl/Texture.java
package com.ryanharter.android.gl;

import static android.opengl.GLES20.GL_TEXTURE_2D;
import static android.opengl.GLES20.glDeleteTextures;
import static android.opengl.GLES20.glGenTextures;

/**
 * Created by rharter on 4/9/14.
 */
public class Texture {

  private int bindUnit = -1;
  private int[] textureId = new int[1];

  public Texture() {
    glGenTextures(1, textureId, 0);
  }

  public void bind(int unit) {
    bindUnit = unit;
    GLState.bindTexture(unit, GL_TEXTURE_2D, textureId[0]);
  }

  public void unbind() {
    if (bindUnit > -1) {
      GLState.bindTexture(bindUnit, GL_TEXTURE_2D, 0);
    }
  }

  @Deprecated public int getTextureId() {
    return getName();
  }

  public int getName() {
    return textureId[0];
  }

  public void destroy() {
    glDeleteTextures(1, textureId, 0);
  }

}

File: gl/src/main/java/com/ryanharter/android/gl/Texture.kt
package com.ryanharter.android.gl

import android.opengl.GLES20.GL_TEXTURE_2D
import android.opengl.GLES20.glDeleteTextures
import android.opengl.GLES20.glGenTextures

/**
 * Created by rharter on 4/9/14.
 */
open class Texture {

  private var bindUnit = -1

  val name: Int

  init {
    val tmp = IntArray(1)
    glGenTextures(1, tmp, 0)
    name = tmp[0]
  }

  open fun bind(unit: Int) {
    bindUnit = unit
    GLState.bindTexture(unit, GL_TEXTURE_2D, name)
  }

  open fun unbind() {
    if (bindUnit > -1) {
      GLState.bindTexture(bindUnit, GL_TEXTURE_2D, 0)
      bindUnit = -1
    }
  }

  open fun destroy() {
    glDeleteTextures(1, intArrayOf(name), 0)
  }

}
